File: scripts/05_kpi5_forecast.py
#scripts/05_kpi5_forecast.py
#eadcount forecasts a 12 meses por departamento

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

#comfigurar carpetas si se ejecuta primero este Script
os.makedirs("../outputs/tablas/05_kpi5_forecast", exist_ok= True)
os.makedirs("../outputs/graficos/05_kpi5_forecast", exist_ok= True)

#cargar datos

df = pd.read_csv("../data/WA_Fn-UseC_-HR-Employee-Attrition.csv")
print(f"Dimensiones: {df.shape}")

# Calcular Turnover por departamento

turnover_dept = df.groupby('Department').agg(
    hc_actual = ('Attrition', 'count'),
    rotaciones =('Attrition', lambda x: (x=='Yes').sum())
).reset_index()
turnover_dept['turnover_anual'] = turnover_dept['rotaciones']/turnover_dept['hc_actual']
turnover_dept['tasa_mensual'] = turnover_dept['turnover_anual']/12

print("\n Turnover por departamento:")
#print(turnover_dept)
print(turnover_dept[['Department','hc_actual','rotaciones','turnover_anual']])

#Parametros de simulacion

fill_rate = 0.95
meses = 12

#proyeccion Mes a Mes

forecast_list = []
for _, row in turnover_dept.iterrows():
    dept = row['Department']
    hc = row['hc_actual']
    tasa_mensual = row['tasa_mensual']

    hc_series = [hc]
    for m in range(1, meses + 1):
        bajas = round(hc_series[-1] * tasa_mensual)
        contrataciones = round(bajas*fill_rate)
        hc_nuevo = hc_series[-1] - bajas + contrataciones
        hc_series.append(hc_nuevo)

        #crear DataFrame para este departemaneto

    df_dept = pd.DataFrame({
        'Department': dept,
        'mes': range(0, meses + 1),
        'headcount': hc_series
    })
    forecast_list.append(df_dept)

df_forecast = pd.concat(forecast_list, ignore_index=True)

df_forecast.to_csv("../outputs/tablas/05_kpi5_forecast/forecast_headcount_Py.csv")
print("\n Tabla de proyeccion guardada en ../outputs/tablas/05_kpi5_forecast/forecast_headcount_Py.csv")

# Grafico de Lineas

plt.figure(figsize = (10 , 6))
colores = {
        'Human Resources': '#C0392B',
        'Research & Development': '#27AE60',
        'Sales': '#1F3C6E'
}
for dept in df_forecast['Department'].unique():
    subset = df_forecast[df_forecast['Department'] == dept]
    plt.plot(subset['mes'], subset['headcount'], marker='o', linewidth=2, 
             label=dept, color=colores.get(dept, '#333333'))

plt.title('Proyección de Headcount a 12 meses por Departamento (fill rate 95%)', fontsize=14)
plt.xlabel('Mes')
plt.ylabel('Número de empleados')
plt.legend(title='Departamento', loc='upper left')
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig("../outputs/graficos/05_kpi5_forecast/forecast_lines_Py.png", dpi=150)
plt.close()
print("✅ Gráfico guardado en ../outputs/graficos/05_kpi5_forecast/forecast_lines_Py.png")

# Resumen Final (headcount a 12 meses)

resumen_final = df_forecast[df_forecast['mes'] == meses].copy()
resumen_final = resumen_final.merge(
    turnover_dept[['Department', 'hc_actual']], on = 'Department', how = 'left'
)
resumen_final['cambio'] = resumen_final['headcount']- resumen_final['hc_actual']

print(f"\n Resumen Final 12 meses:")
print(resumen_final[['Department','hc_actual', 'headcount','cambio']])

resumen_final.to_csv("../outputs/tablas/05_kpi5_forecast/forecast_summary_Py.csv", index = False)
print("Resumen Guardado")

File: scripts/05_kpi5_forecast.R
#scripts/05_kpi5_forecast.R
# Headcount Forecast a 12 meses por departamento.

library(tidyverse)
library(ggplot2)

#Crear las carpetas de output

dir.create("outputs/tablas/05_kpi5_forecast", recursive = TRUE,showWarnings = FALSE)
dir.create("outputs/graficos//05_kpi5_forecast", recursive = TRUE,showWarnings = FALSE)

#cargar datos

df <- read.csv("data/WA_Fn-UseC_-HR-Employee-Attrition.csv")
cat("Dimensiones:", dim(df),"\n")

#Calcular turnover anual por departamento

turnover_dep <- df %>%
  group_by(Department) %>%
  summarise(
    hc_actual = n(),
    rotaciones = sum(Attrition == 'Yes'),
    turnover_anual = rotaciones / hc_actual,
    .groups = 'drop'
)%>%
mutate(tasa_mensual = turnover_anual / 12)
  
cat("\n Turnover por departamento:\n")
print(turnover_dep)

#Parámetros de Simulación

fill_rate <- 0.80 #No tenemos datos de Tasa de cobertura por lo que asumimos un 80% de de capacidad de reclutamiento
meses <- 12

forecast_list <- list()

for (i in 1:nrow(turnover_dep)) {
  depto <- turnover_dep$Department[i]
  hc <- turnover_dep$hc_actual[i]
  tasa_mensual <- turnover_dep$tasa_mensual[i]

#Vector par almacenar headcount mes a mes (mes = 0 Actual)

hc_series <- numeric(meses + 1)
hc_series[1] <- hc

for (m in 1:meses){
  bajas <- round(hc_series[m]* tasa_mensual)
  contrataciones <- round(bajas * fill_rate)
  hc_series[m + 1] <- hc_series[m]-bajas+contrataciones
}

#crear data frame por departamento

df_dept <- tibble(
  Department = depto,
  mes = 0:meses,
  headcount = hc_series
)
  forecast_list[[i]] <- df_dept
}

df_forecast <- bind_rows(forecast_list)

#Exportar tabla de datos

write_csv(df_forecast,"outputs/tablas/05_kpi5_forecast/forecast_headcount_R.csv")
cat("Tabla de proyección guardada en outputs/tablas/05_kpi5_forecast/forecast_headcount_R.csv\n")

# Grafico de lineas

p <- ggplot(df_forecast, aes(x = mes, y = headcount, color = Department, group = Department))+
  geom_line(size = 1.2)+
  geom_point(size = 2)+
  scale_color_manual(values = c("Human Resources" = "#C0392B",
                                "Research & Development" = "#27AE60",
                                "Sales" = "#1F3C6E"))+
  labs(
    title = "Proyección de headcount a  12 meses por Departamento",
    subtitle = paste("Tasa de cobertura (fill rate) =", fill_rate*100,"%"),
    x = "Mes",
    y = "Número de empleados",
    color = "Departamento"
  )+
  theme_minimal()+
  
  theme(legend.position = "botton", legend.box   = "horizontal")

ggsave("outputs/graficos/05_kpi5_forecast/forecast_lines_R.png",p, width = 10, height = 6, dpi = 150)
cat("Gráfico guardado en outputs/graficos/05_kpi5_forecast/forecast_lines_R.png\n")

#Resumen (headcount estimado a 12 meses)

resumen_final <- df_forecast %>%
  filter(mes == meses) %>%
  select(Department, headcount_12m = headcount) %>%
  left_join(turnover_dep %>% select(Department, hc_actual), by = "Department")%>%
  mutate(cambio = headcount_12m - hc_actual)
cat(
  "\n Resumen final a 12 meses"
)
print(resumen_final)

write.csv(resumen_final, "outputs/tablas/05_kpi5_forecast/forecast_summary_R.csv")
cat("Resumen Guardado \n")
